utils/error.hh

```diff
@@ -0,0 +1,92 @@
+/**
+ *		Tempesta FW
+ *
+ * Copyright (C) 2024 Tempesta Technologies, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+#pragma once
+
+#include <array>
+#include <cassert>
+#include <cstdio>
+#include <fmt/format.h>
+#include <source_location>
+#include <sstream>
+#include <string>
+#include <stdexcept>
+
+#include <boost/system/system_error.hpp>
+#include <boost/system/error_code.hpp>
+
+class Exception : public std::runtime_error {
+public:
+	Exception(std::string s) noexcept
+		: std::runtime_error(format_syserr(s))
+	{}
+
+	void
+	print(char *buf, size_t len)
+	{
+		snprintf(buf, len, "%s", what());
+	}
+
+protected:
+	// Add system error code (errno).
+	static std::string
+	format_syserr(std::string msg) noexcept
+	{
+		namespace sys = boost::system;
+
+		auto ec = sys::error_code(errno, sys::system_category());
+		if (ec) {
+			std::stringstream ss;
+			ss << " (" << ec.message() << ", errno=" << ec.value() << ")";
+			msg += ss.str();
+		}
+
+		return msg;
+	}
+};
+
+template <typename... Args>
+class Except : public Exception {
+public:
+	Except() =delete;
+	Except(const Except &) =delete;
+	Except &operator=(const Except &) =delete;
+	~Except() override =default;
+
+	Except(fmt::format_string<Args...> fmt, Args&&... args,
+	       const std::source_location &loc = std::source_location::current()) noexcept
```

Comment:
Good point. Probably we need a macro here

Reply:
Surprisingly, it works!
It shows the right function and line:
`[2024-11-18 15:22:59.646] [access_logger] [error] Test (at tfw_logger.cc:222 in void run_thread(int, int, std::string, std::string, std::string))`

Reply:
@ai-tmpst this is because Except is a template, thus initialization of default `loc` happens in place of the first instantiation of the template for concrete template params pack. This is still an error

Reply:
@Kutumov this also works for me. `std::source_location::current()` is the default value, which is evaluated for a particular function (constructor of instantiated Exception class) call, which may have or not have an arguemnt for `loc`. The code works and I also don't see a problem here...